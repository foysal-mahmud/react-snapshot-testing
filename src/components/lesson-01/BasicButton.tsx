import React from 'react';

interface BasicButtonProps {
  children: React.ReactNode;
  onClick?: () => void;
  disabled?: boolean;
  variant?: 'primary' | 'secondary';
}

const BasicButton: React.FC<BasicButtonProps> = ({ 
  children, 
  onClick, 
  disabled = false,
  variant = 'primary' 
}) => {
  const className = `btn btn-${variant}${disabled ? ' btn-disabled' : ''}`;
  
  return (
    <button 
      className={className}
      onClick={onClick}
      disabled={disabled}
      data-testid="basic-button"
    >
      {children}
    </button>
  );
};

export default BasicButton; 