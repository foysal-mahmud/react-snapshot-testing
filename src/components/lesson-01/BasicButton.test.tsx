import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import BasicButton from './BasicButton';

describe('Lesson 1: BasicButton Component', () => {
  
  // UNIT TESTS - Testing specific behaviors
  describe('Unit Tests - Behavior Testing', () => {
    test('renders children correctly', () => {
      render(<BasicButton>Click me</BasicButton>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveTextContent('Click me');
    });

    test('calls onClick when clicked', () => {
      const mockOnClick = jest.fn();
      render(<BasicButton onClick={mockOnClick}>Click me</BasicButton>);
      
      const button = screen.getByRole('button');
      fireEvent.click(button);
      
      expect(mockOnClick).toHaveBeenCalledTimes(1);
    });

    test('is disabled when disabled prop is true', () => {
      render(<BasicButton disabled>Disabled Button</BasicButton>);
      
      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    test('does not call onClick when disabled', () => {
      const mockOnClick = jest.fn();
      render(
        <BasicButton onClick={mockOnClick} disabled>
          Disabled Button
        </BasicButton>
      );
      
      const button = screen.getByRole('button');
      fireEvent.click(button);
      
      expect(mockOnClick).not.toHaveBeenCalled();
    });

    test('applies correct CSS classes for variants', () => {
      const { rerender } = render(<BasicButton variant="primary">Primary</BasicButton>);
      
      let button = screen.getByRole('button');
      expect(button).toHaveClass('btn', 'btn-primary');
      
      rerender(<BasicButton variant="secondary">Secondary</BasicButton>);
      button = screen.getByRole('button');
      expect(button).toHaveClass('btn', 'btn-secondary');
    });
  });

  // SNAPSHOT TESTS - Testing overall structure
  describe('Snapshot Tests - Structure Testing', () => {
    test('matches snapshot with default props', () => {
      render(<BasicButton>Default Button</BasicButton>);
      const button = screen.getByTestId('basic-button');
      expect(button).toMatchSnapshot();
    });

    test('matches snapshot with primary variant', () => {
      render(<BasicButton variant="primary">Primary Button</BasicButton>);
      const button = screen.getByTestId('basic-button');
      expect(button).toMatchSnapshot();
    });

    test('matches snapshot with secondary variant', () => {
      render(<BasicButton variant="secondary">Secondary Button</BasicButton>);
      const button = screen.getByTestId('basic-button');
      expect(button).toMatchSnapshot();
    });

    test('matches snapshot when disabled', () => {
      render(<BasicButton disabled>Disabled Button</BasicButton>);
      const button = screen.getByTestId('basic-button');
      expect(button).toMatchSnapshot();
    });

    test('matches snapshot with all props', () => {
      render(
        <BasicButton variant="secondary" disabled onClick={() => {}}>
          Complex Button
        </BasicButton>
      );
      const button = screen.getByTestId('basic-button');
      expect(button).toMatchSnapshot();
    });
  });

  // LESSON 1 LEARNING OBJECTIVES
  describe('Lesson 1 Learning Demonstration', () => {
    test('demonstrates finding elements by role', () => {
      render(<BasicButton>Find Me</BasicButton>);
      
      // This is how React Testing Library recommends finding elements
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();
    });

    test('demonstrates finding elements by test id', () => {
      render(<BasicButton>Find Me</BasicButton>);
      
      // Using data-testid is also a good practice
      const button = screen.getByTestId('basic-button');
      expect(button).toBeInTheDocument();
    });

    test('demonstrates testing user interactions', () => {
      const handleClick = jest.fn();
      render(<BasicButton onClick={handleClick}>Interactive Button</BasicButton>);
      
      // Simulate user clicking the button
      const button = screen.getByRole('button');
      fireEvent.click(button);
      
      // Verify the interaction worked
      expect(handleClick).toHaveBeenCalled();
    });
  });
}); 