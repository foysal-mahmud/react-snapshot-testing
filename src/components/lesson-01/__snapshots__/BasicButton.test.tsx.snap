// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lesson 1: BasicButton Component Snapshot Tests - Structure Testing matches snapshot when disabled 1`] = `
<button
  class="btn btn-primary btn-disabled"
  data-testid="basic-button"
  disabled=""
>
  Disabled Button
</button>
`;

exports[`Lesson 1: BasicButton Component Snapshot Tests - Structure Testing matches snapshot with all props 1`] = `
<button
  class="btn btn-secondary btn-disabled"
  data-testid="basic-button"
  disabled=""
>
  Complex Button
</button>
`;

exports[`Lesson 1: BasicButton Component Snapshot Tests - Structure Testing matches snapshot with default props 1`] = `
<button
  class="btn btn-primary"
  data-testid="basic-button"
>
  Default Button
</button>
`;

exports[`Lesson 1: BasicButton Component Snapshot Tests - Structure Testing matches snapshot with primary variant 1`] = `
<button
  class="btn btn-primary"
  data-testid="basic-button"
>
  Primary Button
</button>
`;

exports[`Lesson 1: BasicButton Component Snapshot Tests - Structure Testing matches snapshot with secondary variant 1`] = `
<button
  class="btn btn-secondary"
  data-testid="basic-button"
>
  Secondary Button
</button>
`;
